packages/amplify_core/lib/src/types/common/amplify_auth_provider.dart

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import 'package:amplify_core/amplify_core.dart';
+import 'package:aws_signature_v4/aws_signature_v4.dart';
+
+/// An identifier to use as a key in an [AmplifyAuthProviderRepository] so that
+/// a retrieved auth provider can be typed more accurately.
+class AmplifyAuthProviderToken<T extends AmplifyAuthProvider> extends Token<T> {
+  const AmplifyAuthProviderToken();
+}
+
+abstract class AuthProviderOptions {
+  const AuthProviderOptions();
+}
+
+/// Options required by IAM to sign any given request at runtime.
+class IamAuthProviderOptions extends AuthProviderOptions {
+  const IamAuthProviderOptions({
+    required this.region,
+    required this.service,
+  });
+
+  final String region;
+  final AWSService service;
+}
+
+abstract class AmplifyAuthProvider {
+  const AmplifyAuthProvider();
+
+  Future<AWSBaseHttpRequest> authorizeRequest(
+    AWSBaseHttpRequest request, {
+    covariant AuthProviderOptions? options,
+  });
+}
+
+abstract class AWSIamAmplifyAuthProvider extends AmplifyAuthProvider
+    implements AWSCredentialsProvider {
+  const AWSIamAmplifyAuthProvider();
+
+  @override
+  Future<AWSSignedRequest> authorizeRequest(
+    AWSBaseHttpRequest request, {
+    covariant IamAuthProviderOptions options,
+  });
+}
+
+abstract class TokenAmplifyAuthProvider extends AmplifyAuthProvider {
```

Comment:
on storage branch, Hui defined another abstract class here `TokenIdentityAuthProvider`, which may not be needed for analytics (or API) but if we want to have the most mature version of auth providers in next branch, might be worth checking out this file off storage branch.